### PR TITLE
Rustdoc: Account for const-unstable functions

### LINKED ISSUE #86464: rustdoc rendering for stable-but-const-unstable `const fn` is misleading
I just spent 5min of horror after seeing this rustdoc page:
Screenshot 2021-06-19 at 10-07-35 f32 - Rust

and then I realized that the `const`ness of this function is unstable.  rustdoc should either indicate that, or just not show the `const` at all for stable functions whose `const`ness is still unstable.

Cc @rust-lang/rustdoc @rust-lang/wg-const-eval

### PR DESCRIPTION
**Pull request by fee1-dead**
Fixes #86464

### CODE REVIEW

**Comment by fee1-dead** [comment]
Screenshot:
image

The `unstable` is linked to the tracking issue. Should I change the color?

**Comment by jyn514** [comment]
> The unstable is linked to the tracking issue.

This is awesome, thank you :heart: No, I don't think the color needs to change - I like that it's a different color because it gives you a hint it's a link.

File: src/librustdoc/html/render/mod.rs [hunk 1]
```diff
@@ -888,11 +910,20 @@ fn render_assoc_item(
...
         let defaultness = print_default_space(meth.is_default());
         let abi = print_abi_with_space(header.abi).to_string();
+
+        if !constness.is_empty() {
+            if let Some(ConstStability { level: StabilityLevel::Unstable { .. }, .. }) =
+                meth.const_stability(cx.tcx())
+            {
+                constness = "";
+            }
+        }
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:925]: Why single this out for methods? I would expect this to be consistent wherever rustdoc prints `const`, including for functions. To enforce that you could pass in the const stability to `print_with_space` (by removing the trait implementation and just calling it directly).

File: src/librustdoc/html/render/print_item.rs [hunk 2]
```diff
@@ -430,29 +431,44 @@ fn extra_info_tags(item: &clean::Item, parent: &clean::Item, tcx: TyCtxt<'_>) ->
...
+        + name.as_str().len()
+        + generics_len;
+
     w.write_str("<pre class=\"rust fn\">");
     render_attributes_in_pre(w, it, "");
+    w.reserve(header_len);
     write!(
         w,
-        "{vis}{constness}{asyncness}{unsafety}{abi}fn \
+        "{}{}{}{}{}fn \
```
Comment (jyn514) [inline src/librustdoc/html/render/print_item.rs:464]: Why did you change this? I think it was easier to read before.

File: src/test/rustdoc/const-display.rs [hunk 3]
```diff
@@ -38,10 +39,11 @@ pub const unsafe fn bar_not_gated() -> u32 { 42 }
 pub struct Foo;
 
 impl Foo {
-    // @has 'foo/struct.Foo.html' '//div[@id="method.gated"]/code' 'pub unsafe fn gated() -> u32'
+    // @has 'foo/struct.Foo.html' '//div[@id="method.gated"]/code' 'pub fn gated() -> u32'
+    // @has - '//span[@class="since"]' '1.0.0 (const: unstable)'
     #[stable(feature = "rust1", since = "1.0.0")]
     #[rustc_const_unstable(feature="foo", issue = "none")]
-    pub const unsafe fn gated() -> u32 { 42 }
+    pub const fn gated() -> u32 { 42 }
```
Comment (jyn514) [inline src/test/rustdoc/const-display.rs:46]: Why did this change?

File: src/librustdoc/html/render/print_item.rs (hunk 2 above)
Comment (fee1-dead) [reply src/librustdoc/html/render/print_item.rs:464]: To match `html/render/mod.rs`.

https://github.com/rust-lang/rust/blob/03b845a41f1666654ecffcbaa6a582170ec0ed8d/src/librustdoc/html/render/mod.rs#L898-L934

The length calculation and the printing would reuse most parts instead of creating a temporary string solely to calculate the length. But it might be better to change `{}` into names `{vis}`, etc.

File: src/test/rustdoc/const-display.rs (hunk 3 above)
Comment (fee1-dead) [reply src/test/rustdoc/const-display.rs:46]: In the tests, with `unsafe` it did not render `const` when it is `const_unstable` already. I changed that to test the normal case (a function that was not unsafe). I do not know where the interaction between `unsafe` and `const_unstable` is, but I just wanted to make sure that my changes work well.

File: src/librustdoc/html/render/print_item.rs (hunk 2 above)
Comment (jyn514) [reply src/librustdoc/html/render/print_item.rs:464]: > But it might be better to change {} into names {vis}, etc.

Right, this is what I meant, the rest of the changes are fine :)

File: src/test/rustdoc/const-display.rs (hunk 3 above)
Comment (jyn514) [reply src/test/rustdoc/const-display.rs:46]: Ok. Rather than removing the old test, can you add a new test so that both cases are exercised?

Comment (jyn514) [reply src/test/rustdoc/const-display.rs:46]: Or in other words, it's important to know what the interaction between unsafe and const_unstable is/

File: src/librustdoc/html/render/mod.rs (hunk 1 above)
Comment (fee1-dead) [reply src/librustdoc/html/render/mod.rs:925]: The `Constness` is defined in `rustc_hir`. I am going to add another trait just for `Constness` instead.

Comment (jyn514) [reply src/librustdoc/html/render/mod.rs:925]: Please do not add more traits, they make the code more complicated than necessary. A free function is fine.

Comment (fee1-dead) [reply src/librustdoc/html/render/mod.rs]: Okay, I will change that now.

File: src/librustdoc/html/format.rs [hunk 4]
```diff
@@ -1280,6 +1272,22 @@ impl PrintWithSpace for hir::Mutability {
...
+) -> &'static str {
+    match (c, s) {
+        // const stable or no stability attribute
+        (
+            hir::Constness::Const,
+            Some(ConstStability { level: StabilityLevel::Stable { .. }, .. }),
+        )
+        | (hir::Constness::Const, None) => "const ",
+        // const unstable or not const
+        (hir::Constness::Const, _) | (hir::Constness::NotConst, _) => "",
```
Comment (jyn514) [inline src/librustdoc/html/format.rs:1287]: Listing these patterns doesn't actually filter anything; you can make it a wildcard instead and have the same behavior.
```suggestion
        _ => "",
```

File: src/librustdoc/html/format.rs [hunk 5]
```diff
@@ -1280,6 +1272,22 @@ impl PrintWithSpace for hir::Mutability {
     }
 }
 
+crate fn print_constness_with_space(
+    c: &hir::Constness,
+    s: Option<&ConstStability>,
+) -> &'static str {
+    match (c, s) {
+        // const stable or no stability attribute
```
Comment (jyn514) [inline src/librustdoc/html/format.rs:1280]: Why does this say "or no stability attribute"? Wouldn't that imply it was unstably const?

File: src/test/rustdoc/const-display.rs [hunk 6]
```diff
@@ -7,45 +7,88 @@
...
+// @has - '//span[@class="since"]' '1.0.0 (const: unstable)'
+#[stable(feature = "rust1", since = "1.0.0")]
+#[rustc_const_unstable(feature="foo", issue = "none")]
+pub const unsafe fn foo_unsafe() -> u32 { 42 }
 
 // @has 'foo/fn.foo2.html' '//pre' 'pub const fn foo2() -> u32'
 #[unstable(feature = "humans", issue = "none")]
 pub const fn foo2() -> u32 { 42 }
 
+// @has 'foo/fn.foo2_unsafe.html' '//pre' 'pub const unsafe fn foo2_unsafe() -> u32'
```
Comment (jyn514) [inline src/test/rustdoc/const-display.rs:26]:
```suggestion
// @has 'foo/fn.foo2_unsafe.html' '//pre' 'pub const unsafe fn foo2_unsafe() -> u32'
// @!has - '//span[@class="since"]'
```

File: src/test/rustdoc/const-display.rs [hunk 7]
```diff
@@ -7,45 +7,88 @@
...
+pub const fn foo() -> u32 { 42 }
+
+// @has 'foo/fn.foo_unsafe.html' '//pre' 'pub unsafe fn foo_unsafe() -> u32'
+// @has - '//span[@class="since"]' '1.0.0 (const: unstable)'
+#[stable(feature = "rust1", since = "1.0.0")]
+#[rustc_const_unstable(feature="foo", issue = "none")]
+pub const unsafe fn foo_unsafe() -> u32 { 42 }
 
 // @has 'foo/fn.foo2.html' '//pre' 'pub const fn foo2() -> u32'
 #[unstable(feature = "humans", issue = "none")]
```
Comment (jyn514) [inline src/test/rustdoc/const-display.rs:23]: I know this is an existing test, I hope you don't mind updating it :) feel free to drop this if the test starts failing.
```suggestion
// @!has - '//span[@class="since"]'
#[unstable(feature = "humans", issue = "none")]
```

File: src/test/rustdoc/const-display.rs [hunk 8]
```diff
@@ -7,45 +7,88 @@
...
+
+// @has 'foo/fn.foo2_gated.html' '//pre' 'pub const fn foo2_gated() -> u32'
 #[unstable(feature = "foo2", issue = "none")]
-pub const unsafe fn foo2_gated() -> u32 { 42 }
+pub const fn foo2_gated() -> u32 { 42 }
 
-// @has 'foo/fn.bar2_gated.html' '//pre' 'pub const unsafe fn bar2_gated() -> u32'
+// @has 'foo/fn.foo2_gated_unsafe.html' '//pre' 'pub const unsafe fn foo2_gated_unsafe() -> u32'
+#[unstable(feature = "foo2", issue = "none")]
+pub const unsafe fn foo2_gated_unsafe() -> u32 { 42 }
```
Comment (jyn514) [inline src/test/rustdoc/const-display.rs:48]: You don't need to have an `unsafe` version for each and every new test, just one or two is fine :)

File: src/librustdoc/html/format.rs (hunk 5 above)
Comment (fee1-dead) [reply src/librustdoc/html/format.rs:1280]: it matches this:
```rust
const fn foo() -> u8 { 42 }
```

`const` functions outside the standard library do not have stability attributes.

Comment (jyn514) [reply src/librustdoc/html/format.rs:1280]: Ah ok, you mean if `feature(staged_api)` isn't set. Can you say that here? I was confused at first because I think you can have functions without a stability attribute if it's inherited from the parent module.

File: src/test/rustdoc/const-display.rs [hunk 9]
```diff
@@ -22,26 +30,36 @@ pub const fn foo2() -> u32 { 42 }
 #[rustc_const_stable(feature = "rust1", since = "1.0.0")]
 pub const fn bar2() -> u32 { 42 }
 
-// @has 'foo/fn.foo2_gated.html' '//pre' 'pub const unsafe fn foo2_gated() -> u32'
+
+// @has 'foo/fn.foo2_gated.html' '//pre' 'pub const fn foo2_gated() -> u32'
+// @!has - '//span[@class="since"]'
 #[unstable(feature = "foo2", issue = "none")]
-pub const unsafe fn foo2_gated() -> u32 { 42 }
+pub const fn foo2_gated() -> u32 { 42 }
```
Comment (jyn514) [inline src/test/rustdoc/const-display.rs:37]: So, when you removed the duplicate test, you removed the most complicated one. So the end effect is that this tests less than before your PR.

Sorry to be so picky! This is the last round of review I think :)
```suggestion
pub const unsafe fn foo2_gated() -> u32 { 42 }
```

File: src/test/rustdoc/const-display.rs [hunk 10]
```diff
@@ -22,26 +30,36 @@ pub const fn foo2() -> u32 { 42 }
...
 
-// @has 'foo/fn.bar2_gated.html' '//pre' 'pub const unsafe fn bar2_gated() -> u32'
+// @has 'foo/fn.bar2_gated.html' '//pre' 'pub const fn bar2_gated() -> u32'
 // @has - '//span[@class="since"]' '1.0.0 (const: 1.0.0)'
 #[stable(feature = "rust1", since = "1.0.0")]
 #[rustc_const_stable(feature = "rust1", since = "1.0.0")]
-pub const unsafe fn bar2_gated() -> u32 { 42 }
+pub const fn bar2_gated() -> u32 { 42 }
```
Comment (jyn514) [inline src/test/rustdoc/const-display.rs:43]:
```suggestion
pub const unsafe fn bar2_gated() -> u32 { 42 }
```

File: src/test/rustdoc/const-display.rs [hunk 11]
```diff
@@ -22,26 +30,36 @@ pub const fn foo2() -> u32 { 42 }
...
 
-// @has 'foo/fn.bar_not_gated.html' '//pre' 'pub const unsafe fn bar_not_gated() -> u32'
-pub const unsafe fn bar_not_gated() -> u32 { 42 }
+// @has 'foo/fn.bar_not_gated.html' '//pre' 'pub const fn bar_not_gated() -> u32'
+// @!has - '//span[@class="since"]'
+pub const fn bar_not_gated() -> u32 { 42 }
```
Comment (jyn514) [inline src/test/rustdoc/const-display.rs:47]:
```suggestion
pub const unsafe fn bar_not_gated() -> u32 { 42 }
```

**Comment by jyn514** [comment]
@bors r+

Thanks for the fix!
